File: pkg/utils/apikey/apikey.go
// Copyright 2026 ScitiX
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package apikey

import (
	"crypto/subtle"
	"strings"
)

const (
	// RoleAdmin is the admin role string.
	RoleAdmin = "admin"
	// RoleTenant is the tenant role string.
	RoleTenant = "tenant"
)

// AdminKeyManager handles constant-time comparison of the admin API key.
type AdminKeyManager struct {
	adminKey string
}

// NewAdminKeyManager creates a new AdminKeyManager.
// adminKey is the raw admin key string. When empty, IsAdminKey always returns false.
func NewAdminKeyManager(adminKey string) *AdminKeyManager {
	return &AdminKeyManager{adminKey: adminKey}
}

// IsAdminKey reports whether the provided key equals the admin key using
// constant-time comparison to prevent timing attacks.
func (m *AdminKeyManager) IsAdminKey(key string) bool {
	if m.adminKey == "" {
		return false
	}
	return subtle.ConstantTimeCompare(
		[]byte(strings.TrimSpace(key)),
		[]byte(m.adminKey),
	) == 1
}

File: dashboard/lib/queries/apikey.ts
/**
 * Copyright 2026 ScitiX
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Query options and mutations for API Key resources (admin, per-cluster)

import { useQueryClient } from "@tanstack/react-query"
import { currentApiClient } from "@/lib/api/client"
import { delayedInvalidate } from "./utils"

// ─── Query options ─────────────────────────────────────────────────────────────

export const apiKeysQueryOptions = (namespace?: string) =>
  currentApiClient().queryOptions(
    "get",
    "/admin/api-keys",
    {
      params: { query: namespace ? { namespace } : undefined },
    },
    {
      // ListAPIKeysResult is a bare APIKeyItem[] (the endpoint returns the
      // full list without pagination), not an envelope with `.items`.
      select: (data) => data ?? [],
    },
  )

// ─── Admin mutations ───────────────────────────────────────────────────────────

export function useCreateApiKey() {
  const qc = useQueryClient()
  return currentApiClient().useMutation("post", "/admin/api-keys", {
    onSuccess: () => delayedInvalidate(qc, ["get", "/admin/api-keys"]),
  })
}

export function useDeleteApiKey() {
  const qc = useQueryClient()
  return currentApiClient().useMutation("delete", "/admin/api-keys/{name}", {
    onSuccess: () => delayedInvalidate(qc, ["get", "/admin/api-keys"]),
  })
}

export function usePromoteApiKey() {
  const qc = useQueryClient()
  return currentApiClient().useMutation("post", "/admin/api-keys/{name}/promote", {
    onSuccess: () => delayedInvalidate(qc, ["get", "/admin/api-keys"]),
  })
}

// ─── Key selection ─────────────────────────────────────────────────────────────

/**
 * The key a browser-side caller should authenticate an E2B request with: the
 * newest one whose plaintext the platform still holds.
 *
 * The E2B surface takes API keys only — never the session JWT — so anything the
 * console drives through it has to pick a key on the user's behalf. Legacy keys
 * are stored as a hash alone and cannot be sent upstream, so they are not
 * candidates however recent they are.
 */
export function pickUsableApiKey<T extends { rawToken?: string; issuedAt?: string }>(
  keys: T[] | undefined,
): T | undefined {
  return (keys ?? [])
    .filter((k) => !!k.rawToken)
    .sort((a, b) => (b.issuedAt ?? "").localeCompare(a.issuedAt ?? ""))
    .at(0)
}
